test/functional/feature_taproot.py

```diff
@@ -0,0 +1,600 @@
+            in_value = sum(utxo.output.nValue for utxo in input_utxos) - fee
+            tx.vin = [CTxIn(outpoint = input_utxos[i].input, nSequence = random.randint(min_sequence, 0xffffffff)) for i in range(inputs)]
+            tx.wit.vtxinwit = [CTxInWitness() for i in range(inputs)]
+            self.log.info("Test: %s" % (", ".join(utxo.spender.comment for utxo in input_utxos)))
+
+            # Add 1 to 4 outputs
+            outputs = random.choice([1,2,3,4])
+            assert in_value >= 0 and fee - outputs * DUST_LIMIT >= MIN_FEE
+            for i in range(outputs):
+                tx.vout.append(CTxOut())
+                if in_value <= DUST_LIMIT:
+                    tx.vout[-1].nValue = DUST_LIMIT
+                elif i < outputs - 1:
+                    tx.vout[-1].nValue = in_value
+                else:
+                    tx.vout[-1].nValue = random.randint(DUST_LIMIT, in_value)
+                in_value -= tx.vout[-1].nValue
+                tx.vout[-1].scriptPubKey = random.choice(host_spks)
+            fee += in_value
+            assert(fee >= 0)
+
+            # For each inputs, make it fail once; then succeed once
+            for fail_input in range(inputs + 1):
+                # Wipe scriptSig/witness
+                for i in range(inputs):
+                    tx.vin[i].scriptSig = CScript()
+                    tx.wit.vtxinwit[i] = CTxInWitness()
+                # Fill inputs/witnesses
+                for i in range(inputs):
+                    fn = input_utxos[i].spender.sat_function
+                    fn(tx, i, [utxo.output for utxo in input_utxos], i != fail_input)
+                # Submit to mempool to check standardness
+                standard = fail_input == inputs and all(utxo.spender.is_standard for utxo in input_utxos) and tx.nVersion >= 1 and tx.nVersion <= 2
+                if standard:
+                    self.nodes[0].sendrawtransaction(tx.serialize().hex(), 0)
+                    assert(self.nodes[0].getmempoolentry(tx.hash) is not None)
+                else:
+                    assert_raises_rpc_error(-26, None, self.nodes[0].sendrawtransaction, tx.serialize().hex(), 0)
+                # Submit in a block
+                tx.rehash()
+                msg = ','.join(utxo.spender.comment + ("*" if n == fail_input else "") for n, utxo in enumerate(input_utxos))
+                self.block_submit(self.nodes[0], [tx], msg, witness=True, accept=fail_input == inputs, cb_pubkey=random.choice(host_pubkeys), fees=fee)
+
+    def build_spenders(self):
+        VALID_SIGHASHES = [0,1,2,3,0x81,0x82,0x83]
+        spenders = []
+
+        for annex in [None, bytes([ANNEX_TAG]) + random_bytes(random.randrange(0, 250))]:
+            standard = annex is None
+            sec1, sec2 = ECKey(), ECKey()
+            sec1.generate()
+            sec2.generate()
+            pub1, pub2 = sec1.get_pubkey(), sec2.get_pubkey()
```

Comment:
Should these be updated to be BIP schnorr pubkeys (32-byte arrays)?

Reply:
Done by completely rewriting the Schnorr Python code.